torch/ao/quantization/experimental/_gpu_quantization/quant_primitives.py

```diff
@@ -0,0 +1,384 @@
+            input.device.type
+        )
+
+    # cublas paths
+    if not mat2.is_contiguous():  # silently gives incorrect result without this
+        mat2 = mat2.contiguous()
+    if (not input.is_contiguous()) and (
+        input.shape[0] % 8 != 0
+    ):  # gives cryptic error without this
+        input = (
+            input.contiguous()
+        )  # (it seems the transpose makes cublas check the above j constraint on i)
+    return out_dtype(torch.ops.aten.mm.default, torch.int32, input, mat2)
+
+
+# copy-pasta of https://www.internalfb.com/intern/anp/view/?id=3350736
+def dynamically_quantize_per_tensor(
+    x,
+    quant_min,
+    quant_max,
+    target_dtype,
+    qscheme=torch.per_tensor_affine,  # for now, reuse existing qscheme enum
+):
+    # assumes affine quantization
+
+    # default setup for affine quantization of activations
+    eps = torch.finfo(torch.float32).eps
+
+    if qscheme == torch.per_tensor_affine:
+        # get min and max
+        # TODO(future): make torch.aminmax work on cpu-half
+        # min_val, max_val = torch.aminmax(x)
+        min_val = torch.min(x)
+        max_val = torch.max(x)
+
+        # calculate scale and zero point based on min and max
+        # reference: https://fburl.com/code/srbiybme
+        min_val_neg = torch.min(min_val, torch.zeros_like(min_val))
+        max_val_pos = torch.max(max_val, torch.zeros_like(max_val))
+        device = min_val_neg.device
+
+        scale = (max_val_pos - min_val_neg) / float(quant_max - quant_min)
+        # TODO(future): make torch.clamp with scalar work on cpu-half
+        scale = torch.clamp(scale, min=eps).reshape(1)
+        zero_point = quant_min - torch.round(min_val_neg / scale).to(torch.int)
+        zero_point = torch.clamp(zero_point, quant_min, quant_max)
+
+        # quantize based on qmin/qmax/scale/zp
+        # reference: torch/ao/quantization/fx/_decomposed.py?lines=63
+        quant = torch.clamp(
+            torch.round(x / scale) + zero_point, quant_min, quant_max
+        ).to(target_dtype)
+
+    else:
+        assert qscheme == torch.per_tensor_symmetric, f"unsupported qscheme {qscheme}"
+        # assert quant_min == -1 * quant_max, "unsupported quant_min/quant_max"
+        amax = torch.max(torch.abs(x))
+        scale = amax / (float(quant_max - quant_min) / 2)
+        scale = torch.clamp(scale, min=eps).reshape(1)
+        quant = torch.clamp(torch.round(x / scale), quant_min, quant_max).to(
+            target_dtype
+        )
+        # do not create a tensor for zero_point as this is expensive
+        zero_point = None
+
+    return quant, scale, zero_point
+
+
+# taken from
+# https://github.com/mit-han-lab/smoothquant/blob/2f87951dacfb9238d8d657f52ae83a82a3c9ba0c/smoothquant/fake_quant.py#L26
+# and slightly modified
+def quantize_activation_per_token_absmax(t):
+    n_bits = 8
+    # if the shape of t is [B, N, K], the shape of scales will be [B, N, 1]
+
+    scales = t.abs().max(dim=-1, keepdim=True)[0]
```

Comment:
should this instead use `amax` directly: `t.abs().amax(dim=-1, keepdim=True)`?

Reply:
yeah i can change that